electrum/json_db.py

```diff
@@ -53,6 +57,122 @@ class TxFeesValue(NamedTuple):
     num_inputs: Optional[int] = None
 
 
+_RaiseKeyError = object() # singleton for no-default behavior
+
+class StorageDict(dict):
+
+    def __init__(self, data, db, path):
+        self.db = db
+        self.path = path
+        # recursively convert dicts to storagedict
+        for k, v in list(data.items()):
+            self.__setitem__(k, v)
+
+    def convert_key(self, key):
+        # convert int, HTLCOwner to str
+        return str(int(key)) if isinstance(key, int) else key
+
+    def __setitem__(self, key, v):
```

Comment:
storage.write() takes db.lock;
should StorageDict.__setitem__ and similar methods not take db.lock?